### ISSUE TITLE
Proposal: make autoconf, automake, bintools, libtool virtual packages to fix reports of darwin libtool breakage

### ISSUE THREAD

**Issue by goxberry** [spack/spack#7172]
On darwin, spack is sort of a disaster because it treats darwin like it's GNU Linux and builds a GNU toolchain, but the OS is BSD-based with some GPLv2 stuff thrown in here or there, and differences between some of the BSD and GNU tools cause linking problems.

Spack's first attempt at constructing a build toolchain was to go the all-GNU route. Let's call this Version 1.

I attempted to fix this in #7060 because it broke `subversion`. Mistakes notwithstanding, the idea was to symlink `libtool` to `glibtool` and `libtoolize` to `glibtoolize`. This fixes `subversion`, but then breaks a number of other packages. Let's call this Version 2.

In addition to the breakage caused by #7060, I attempted to fix this again via #7073, which just installs `glibtool` and `glibtoolize`. This satisfies build systems that detect Darwin and explicitly look for the GNU tools in these paths (`subversion`, among others). It also satisfies any build system that detects Apple's BSD `libtool` on its own, but it breaks any `AutotoolsPackage` that calls `autoconf` and expects to find `libtool` in the `binutils` prefix, notably `gcc`. (See #7140, #7163.) Let's call this Version 3.

It seems to me like a better way forward would look like:

[ ] Make a `cctools` package from https://opensource.apple.com/source/cctools/cctools-895/, using [MacPorts](https://github.com/macports/macports-ports/blob/master/devel/cctools/Portfile) and [Homebrew](https://github.com/Homebrew/homebrew-core/blob/master/Formula/cctools.rb) packages as guides. This package would install Apple's toolchain into the spack tree because if spack and OpenGL have taught me anything, it's that forcing users to configure external packages is a nightmare, and having a package that just reinstalls system utilities is way less painful.

[ ] Decide how spack is going to distinguish between BSD and GNU versions of the above in the 
"version" field of a spec of virtual packages providing `autoconf`, etc. Some build systems seem unaffected by using either version. Other build systems crash and burn if the wrong one is used. This distinction determines how to enforce the BSD vs GNU constraint during concretization.

[ ] Rename existing GNU build tools: e.g., `autoconf` -> `gnu-autoconf`

[ ] Make `autoconf`, etc., virtual packages

[ ] Add `provides` directives to `cctools`, GNU build tools, and other any BSD versions of build tools around (`bmake` is one, I believe), using the versioning scheme decided above

[ ] Profit.

Given the amount of work involved, and the fact that it's going to affect a lot of stuff, I figured it might be a good idea to run this by some of the core devs first before doing anything.

@tgamblin @alalazo @davydden @adamjstewart

**Comment by goxberry** [comment]
I'll be working on the `cctools` part tonight. If there's reasonable consensus on this being a reasonable path forward, the other stuff seems like it could be done in a couple days.

As for versioning between `gnu` and `bsd`, I'd say using `gnu` and `bsd` version tags would be fine, or maybe `2.69-gnu` versus `1.3-bsd`, following the convention established by Apple Clang. I am open to other suggestions.

**Comment by tgamblin** [comment]
I like this idea -- especially if it makes it easier to build things on macs with Spack.

Questions/comments:

1. The `cctools` package would just repackage Apple's stuff and make it provide all the virtual dependencies?  I believe that should work ok.
2. I have thought for a while that it might be a good idea to allow regular packages to share names with virtual packages, provided that the regular package actually provides the virtual package of the same name.  That would keep the hashes stable.  I think this might be more intuitive for something like `pkgconf` and `pkg-config`, where we had to introduce a fake name to make them provide the same virtual.  I think this might help here.  If you're up for looking at it.
3. Do we need version suffixes at all?  Is that intended to allow someone to explicitly depend on the GNU or BSD versions?  What if they all also provided `gnu-*` virtual dependencies?

**Comment by goxberry** [comment]
@tgamblin 

> The cctools package would just repackage Apple's stuff and make it provide all the virtual dependencies.

Yes, this is what MacPorts does for gcc, and MacPorts is closer philosophically to what spack is trying to do.

> I have thought for a while that it might be a good idea to allow regular packages to share names with virtual packages, provided that the regular package actually provides the virtual package of the same name. That would keep the hashes stable. I think this might be more intuitive for something like pkgconf and pkg-config, where we had to introduce a fake name to make them provide the same virtual. I think this might help here. If you're up for looking at it.

I can take a look at it. My Python skills are middling -- I'm learning a ton from this project, particularly about decorators -- but if I can figure it out, happy to give it a shot.

> Do we need version suffixes at all? Is that intended to allow someone to explicitly depend on the GNU or BSD versions? What if they all also provided gnu-* virtual dependencies?

Yes, that's intended to allow someone to explicitly select the GNU or BSD version if the build system won't work with other versions. I'm not totally sure why this happens, but the problems I see usually have to do with GNU `ar` creating binaries that lack some sort of architecture specifier that causes `ld` to throw a warning (`ld: warning: file was built for archive which is not the architecture being linked (x86_64)`). Googling around wasn't particularly helpful, so I ended up finding the fix of switching `ar` versions through trial and error.

For a particular example, I know `subversion` looks for GNU libtool as `glibtoolize`. I've run into the issue with the `ld` warnings on a few packages, but I've submitted so many pull requests lately I can't remember the particulars (this is not meant as a humblebrag).

**Comment by goxberry** [comment]
After more reading, I understand a little better what's going on (and realize how much I still don't understand).

Part of what's happening is that `AutotoolsPackage` is calling `libtoolize` in the `autoreconf` stage. GMP forces `autoreconf`, which is why I can reproduce the error in #7140 when trying to install `gmp`. I tried to set a conditional to get `AutotoolsPackage` to call `glibtoolize` when `sys.platform == 'darwin'` instead, but it didn't work. So I went back to the symlinking scheme in #7060, but fixed it with correct symlinking in #7179. I think this will make `libtool` less of an issue.

The weird linking errors I see on darwin seem like they come from using `ar` from `binutils`, which is a non-system `ar`. MacPorts seems to deal with this issue by having both `apple-cctools` and `binutils` packages, although they call the former package `cctools`. (Spack already has a package named `cctools`, so I had to add the `apple-` prefix.) Most packages in MacPorts are built with `apple-cctools` if they require `binutils`-like utilities (`gcc` seems to be a good example); Homebrew doesn't seem to need `apple-cctools` unless the system is sufficiently old (10.6 or older, I think). MacPorts also reinstalls the Apple system linker `ld64`; Homebrew doesn't unless the system is sufficiently old (also 10.6 or older, I think).

So now I think it's probably not a good idea to have `apple-cctools` provide `binutils` because each has different binaries -- it's not simply an issue of GNU versus BSD versions, although that's true for a few commands common to each package. Prefixing the `binutils` binaries with `g` (that is, `nm` becomes `gnm`) is done on Homebrew, and packages in Homebrew depending on `binutils` seem to build just fine; this prefixing eliminates namespace collisions with `apple-cctools` binaries on darwin.
